cpp/src/arrow/testing/random.cc

```diff
@@ -182,6 +188,9 @@ static std::shared_ptr<NumericArray<ArrowType>> GenerateNumericArray(int64_t siz
 
   buffers[1] = *AllocateBuffer(sizeof(CType) * size);
   options.GenerateData(buffers[1]->mutable_data(), size);
+  if (std::is_same<ArrowType, Date64Type>::value) {
+    GenerateFullDayMillisNoNan(buffers[1]->mutable_data(), size);
+  }
```

Comment:
This is intentional. Since we need to generate multiples of 86400000 for `Date64Type`, random numbers are generated in `GenerateData` and ultimately multiplied by 86400000 in [`GenerateFullDayMillisNoNan`](https://github.com/JabariBooker/arrow/blob/ARROW-10924/cpp/src/arrow/testing/random.cc#L172). The behavior of `GenerateFullDayMillisNoNan` is separated from `GenerateOptions` since this class doesn't know the arrow type, only the underlying C type. However, behavior of `GenerateFullDayMillisNoNan` could be moved into the above `if` statement if needed.